discord/bot.py

```diff
@@ -191,270 +198,583 @@ def get_application_command(
+            command registration behavior.
 
         .. versionadded:: 2.0
+
+
+        Parameters
+        ----------
+        guild_id: Optional[:class:`int`]
+            The guild id to get the desynced commands for, else global commands if unspecified.
+
+        Returns
+        -------
+        List[Dict[str, Any]]
+            A list of the desynced commands. Each will come with at least the ``cmd`` and ``action`` keys, which
+            respectively contain the command and the action to perform. Other keys may also be present depending on
+            the action, including ``id``.
         """
-        # TODO: Write this function as described in the docstring (bob will do this)
-        raise NotImplementedError
+        # We can suggest the user to upsert, edit, delete, or bulk upsert the commands
+
+        return_value = []
+        cmds = copy.deepcopy(self.pending_application_commands)
 
-    async def register_commands(self) -> None:
+        if guild_id is None:
+            registered_commands = await self.http.get_global_commands(self.user.id)
+            pending = [cmd for cmd in cmds if cmd.guild_ids is None]
+        else:
+            registered_commands = await self.http.get_guild_commands(self.user.id, guild_id)
+            pending = [cmd for cmd in cmds if cmd.guild_ids is not None and guild_id in cmd.guild_ids]
+
+        registered_commands_dict = {cmd["name"]: cmd for cmd in registered_commands}
+        to_check = {
+            "default_permission": None,
+            "name": None,
+            "description": None,
+            "options": [
+                "type",
+                "name",
+                "description",
+                "autocomplete",
+                "choices"
+            ]
+        }
+        # First let's check if the commands we have locally are the same as the ones on discord
+        for cmd in pending:
+            match = registered_commands_dict.get(cmd.name)
+            if match is None:
+                # We don't have this command registered
+                return_value.append({
+                    "command": cmd,
+                    "action": "upsert"
+                })
+                continue
+
+            as_dict = cmd.to_dict()
+
+            for check in to_check:
+                if type(to_check[check]) == list:
+                    for opt in to_check[check]:
+
+                        cmd_vals = [val.get(opt, MISSING) for val in as_dict[check]]
+                        for i, val in enumerate(cmd_vals):
+                            # We need to do some falsy conversion here
+                            # The API considers False (autocomplete) and [] (choices) to be falsy values
+                            falsy_vals = (False, [])
+                            if val in falsy_vals:
+                                cmd_vals[i] = MISSING
+                        match_vals = [val.get(opt, MISSING) for val in match[check]]
```

Comment:
I got a `KeyError: 'options'` here. With `{'id': '938247032953511967', 'application_id': '443303689411887104', 'version': '938247032953511968', 'default_permission': False, 'default_member_permissions': None, 'type': 1, 'name': 'shutdown', 'description': 'Shutdown the bot.', 'dm_permission': None}` as `match`

Reply:
resolved in 905f6e4301db725cdb231890218a9df2df035fc0